### DISCUSSION TITLE
Something akin to Clojure's `defonce`

### DISCUSSION THREAD

**Discussion by saikyun** [janet-lang/janet#607] [Q&A] [answered]
When repling, I find it very practical to do things similar to:

```
(defonce state @{:cats 0})

(update state :cats inc)
```

Then just eval the whole thing. But I'd like to do so without having `state` being redefined. I couldn't figure out a way to do this. Perhaps putting things into the env? But would the compiler accept that?

**Comment by pyrmont** [comment, marked as answer]
How about:

```clojure
(defmacro defonce
  "Define a value once."
  [name & more]
  (if (= nil (dyn name))
    ~(def ,name ,;more)))
```

**Reply by saikyun** [reply]
Thanks, that's great! :)

**Reply by saikyun** [reply]
Modified it ever so slightly:
```clojure
(defmacro defonce
  "Define a value once."
  [name & more]
  (when (nil? (dyn name))
    ~(def ,name ,;more)))
```

**Reply by pyrmont** [reply]
Ah yeah, sorry, I wrote it the way I did so that it could slot into `boot.janet` maybe :P (`when` and `nil` aren't set at the top of the file). For your own code, yours is better.

**Reply by saikyun** [reply]
Aha, that's an interesting point. I hadn't thought of that. Thanks for the added context. :)
